jina/types/arrays/memmap.py

```diff
@@ -154,32 +154,52 @@ def append(self, doc: 'Document', flush: bool = True) -> None:
             self._header.flush()
             self._body.flush()
 
-    def __getitem__(self, key: Union[int, str]) -> 'Document':
+    def _iteridx_by_slice(self, s: slice):
+        start, stop, step = s.start or 0, s.stop or self.__len__(), s.step or 1
+        if (
+            start <= -self.__len__()
+            or start >= self.__len__()
+            or stop <= -self.__len__()
```

Comment:
LGTM overall, just a nit/question:

In python list implementation, slicing with out of bounds indices is allowed, for instance:
```
>>> a = [1,2,3,4,5,6,7]
>>> a[0:-6]
[1]
>>> a[0:-7]
[]
>>> a[0:-8]
[]
>>> a[0:-9]
[]
>>> a[5:]
[6, 7]
>>> a[8:]
[]
>>> a[8:-10]
[]
```
Should we remove this constraint to align with the python implementation?

Reply:
I've added support for out of bound slices + some tests. lmk if it's okay now

Reply:
Sure, LGTM!